### PR TITLE
Introduce a minimum partition version

### PR DESCRIPTION
**Pull request by dominiklohmann**
This introduces a minimum partition version for VAST in order to enable future cleanup of things we know only existed in databases where all partitions have a partition version lower than the minimum. This PR also bumps the minimum partition version to 1.

### CODE REVIEW

File: libvast/src/system/catalog.cpp [hunk 1]
```diff
@@ -407,9 +407,29 @@ catalog(catalog_actor::stateful_pointer<catalog_state> self,
...
-      atom::merge,
-      std::shared_ptr<std::map<uuid, partition_synopsis_ptr>>& ps) -> atom::ok {
+    [=](atom::merge,
+        std::shared_ptr<std::map<uuid, partition_synopsis_ptr>>& ps)
+      -> caf::result<atom::ok> {
+      auto unsupported_partitions = std::vector<uuid>{};
+      for (const auto& [uuid, synopsis] : *ps) {
+        auto [min, max] = version::supported_vast_version_for_partition_version(
+          synopsis->version);
+        if (max)
```
Comment (lava) [inline libvast/src/system/catalog.cpp:417]: This should probably be something like `if (max < current_partition_version)`

**Review by lava — APPROVED** [review]
Useful mechanism to have in general

File: libvast/src/config.cpp.in [hunk 2]
```diff
@@ -15,7 +16,22 @@ const char* major = "@VAST_VERSION_MAJOR@";
...
 const char* tweak = "@VAST_VERSION_TWEAK@";
-const uint64_t partition_version = @VAST_PARTITION_VERSION@;
+const uint64_t current_partition_version = @VAST_PARTITION_VERSION@;
+
+std::pair<std::string_view, std::optional<std::string_view>>
+supported_vast_version_for_partition_version(
+  uint64_t partition_version) noexcept {
+  switch (partition_version) {
+    case 0:
+      return {"VAST v1.0", "VAST v2.4"};
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/src/config.cpp.in:26]: This is the PR that breaks it by introducing a minimum of 1.

File: libvast/src/system/catalog.cpp (hunk 1 above)
Comment (dominiklohmann) [reply libvast/src/system/catalog.cpp:417]: Max is a VAST version, it cannot be compared with a partition version. That doesn't make sense to me.

File: libvast/src/config.cpp.in [hunk 3]
```diff
@@ -15,7 +16,22 @@ const char* major = "@VAST_VERSION_MAJOR@";
 const char* minor = "@VAST_VERSION_MINOR@";
 const char* patch = "@VAST_VERSION_PATCH@";
 const char* tweak = "@VAST_VERSION_TWEAK@";
-const uint64_t partition_version = @VAST_PARTITION_VERSION@;
+const uint64_t current_partition_version = @VAST_PARTITION_VERSION@;
+
+std::pair<std::string_view, std::optional<std::string_view>>
+supported_vast_version_for_partition_version(
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/src/config.cpp.in:22]: The original idea was simply for it to be in the same place as the partition version, and that has to be in here.

File: libvast/src/system/catalog.cpp (hunk 1 above)
Comment (lava) [reply libvast/src/system/catalog.cpp:551]: Ah, so if the second version is not null that already implies that the partition version is currently unsupported? This got me confused, so I'd suggest changing the names to `auto [version_introduced, version_eol] = ...` or something like that.

Comment (dominiklohmann) [reply libvast/src/system/catalog.cpp:551]: I'll just make that a struct that has two members named `introduced` and `eol`, that's probably for the better.

File: libvast/src/config.cpp.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (lava) [reply libvast/src/config.cpp.in]: The disadvantage is that it's not a C++ file, so it doesn't get the autocomplete and clang-tidy support (at least on my system).

Also, while browsing around for a good candidate I noticed that we already have another breaking-change mechanism in `db_version.hpp`, does it make sense to integrate these (eg. by saying db version X implies min partition version Y) or do you want to keep them separate?

[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/src/config.cpp.in]: It does generate into a C++ file during build, so your clangd should be picking up the generated file in the build directory. It does for me, at least.

I don't think we ever fully integrated the old db_version, as it didn't have an upgrade path built in like the partition version. Should I remove it as it's probably obsolete?

[the remark this answers is not in the document]
Comment (lava) [reply libvast/src/config.cpp.in]: In principle I think the location of the db_version check is better, i.e. when spawning the node and not only after the old state was loaded from disk (when we further disentangle import and export in the future, VAST may have already been running for a while before it notices the invalid partition)

But it's not obvious how to implement the semantics for that: One direction is easy (e.g. db_version x => minimum partition version 1), but it's not clear when we can safely increase the db version as that would probably require a backchannel from the catalog.

So it's probably fine to leave it for now and worry  about it later.

**Pull request merged**